lib/inventory/metadata/metadata.go

```diff
@@ -225,25 +221,45 @@ func (c *fetchConfig) fetchContainerOrchestrator() string {
 
 // fetchCloudEnvironment returns aws, gpc or azure if the instance is running on
 // such cloud environments.
-func (c *fetchConfig) fetchCloudEnvironment() string {
-	if c.awsHTTPGetSuccess() {
-		return "aws"
-	}
-	if c.gcpHTTPGetSuccess() {
-		return "gcp"
-	}
-	if c.azureHTTPGetSuccess() {
-		return "azure"
+func (c *fetchConfig) fetchCloudEnvironment(ctx context.Context) string {
+	ctx, cancel := context.WithTimeout(ctx, 3*time.Second)
+	defer cancel()
+
+	// kick off 3 checks in parallel, at most 1 will succeed
+	cloudEnv := make(chan string, 3)
+
+	go func() {
+		if c.awsHTTPGetSuccess(ctx) {
+			cloudEnv <- "aws"
+		}
+	}()
+
+	go func() {
+		if c.gcpHTTPGetSuccess(ctx) {
+			cloudEnv <- "gcp"
+		}
+	}()
+
+	go func() {
+		if c.azureHTTPGetSuccess(ctx) {
+			cloudEnv <- "azure"
+		}
+	}()
+
+	select {
+	case env := <-cloudEnv:
+		return env
+	case <-ctx.Done():
+		return ""
 	}
```

Comment:
This now takes the full 3 seconds if not running in any cloud. Is that not a common case and won't it be much slower than the current code most of the time?

Reply:
Yep, really good point. Maybe each fetch should unconditionally write to the channel and we return as soon as we get a non-empty string, 3 empty strings, or the context expires (whichever is soonest).